Mathlib/AlgebraicGeometry/EllipticCurve/DivisionPolynomial.lean

```diff
@@ -0,0 +1,423 @@
+/-
+Copyright (c) 2023 David Kurniadi Angdinata. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Kurniadi Angdinata
+-/
+import Mathlib.AlgebraicGeometry.EllipticCurve.Affine
+import Mathlib.Data.Int.Parity
+import Mathlib.NumberTheory.EllipticDivisibilitySequence
+
+/-!
+# Division polynomials of Weierstrass curves
+
+This file defines certain polynomials associated to division polynomials of Weierstrass curves and
+computes their leading terms. These are defined in terms of the auxiliary sequences of normalised
+elliptic divisibility sequences defined in `Mathlib.NumberTheory.EllipticDivisibilitySequence`.
+
+## Mathematical background
+
+Let $W$ be a Weierstrass curve over a commutative ring $R$. The sequence of $n$-division polynomials
+$\psi_n \in R[X, Y]$ of $W$ is the normalised elliptic divisibility sequence with initial values
+ * $\psi_0 := 0$,
+ * $\psi_1 := 1$,
+ * $\psi_2 := 2Y + a_1X + a_3$,
+ * $\psi_3 := 3X^4 + b_2X^3 + 3b_4X ^ 2 + 3b_6X + b_8$, and
+ * $\psi_4 := \psi_2 \cdot
+    (2X^6 + b_2X^5 + 5b_4X^4 + 10b_6X^3 + 10b_8X^2 + (b_2b_8 - b_4b_6)X + (b_4b_8 - b_6^2))$.
+
+Furthermore, define the associated sequences $\phi_n, \omega_n \in R[X, Y]$ by
+ * $\phi_n := X\psi_n^2 - \psi_{n + 1}\psi_{n - 1}$, and
+ * $\omega_n := \tfrac{1}{2\psi_n} \cdot (\psi_{2n} - \psi_n^2(a_1\phi_n + a_3\psi_n^2)$.
```

Comment:
I'll fix the definition and add a short note explaining the issue with division by 2. I think it's a great idea to have multiple files since they're getting quite long. I'll propagate this to my downstream PRs.

I think we don't want to define $$\omega$$ in this PR (presumably that's what you mean by including #13057 since it's now broken to a few smaller PRs), because that is massive on its own.

Reply:
I added a very short description of how to define $\omega$ mathematically at the module docstrings. The actual implementation details under `Implementation Notes` we can add when we actually PR the definition of $\omega$.